internal/validitywindow/client.go

```
@@ -0,0 +1,122 @@
+) *BlockFetcherClient[B] {
+	return &BlockFetcherClient[B]{
+		syncClient: typedclient.NewSyncTypedClient(baseClient, &blockFetcherMarshaler{}),
+		parser:     parser,
+		sampler:    sampler,
+	}
+}
+
+// FetchBlocks fetches blocks from peers **backward**.
+//   - It stops when `minTS` is met (this can update dynamically, e.g., via `UpdateSyncTarget`).
+//   - Each request is limited by the node's max execution time (currently ~50ms),
+//     meaning multiple requests may be needed to retrieve all required blocks.
+//   - If a peer is unresponsive or sends bad data, we retry with another
+func (c *BlockFetcherClient[B]) FetchBlocks(ctx context.Context, blk Block, minTimestamp *atomic.Int64) <-chan FetchResult[B] {
+	resultChan := make(chan FetchResult[B], 1)
+
+	go func() {
+		c.lastBlock = blk
+		req := &BlockFetchRequest{MinTimestamp: minTimestamp.Load(), BlockHeight: blk.GetHeight() - 1}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Multiple blocks can share the same timestamp, so we have not filled the validity window
+			// until we find and include the first block whose timestamp is strictly less than the minimum
+			// timestamp. This ensures we have a complete and verifiable validity window
+			if c.lastBlock.GetTimestamp() < minTimestamp.Load() {
+				close(resultChan)
+				return
+			}
+
+			nodeID, ok := c.sampleNodeID(ctx)
+			if !ok {
+				time.Sleep(backoff)
+				continue
+			}
+
+			reqCtx, cancel := context.WithTimeout(ctx, requestTimeout)
+			response, err := c.syncClient.SyncAppRequest(reqCtx, nodeID, req)
+			cancel()
+
+			if err != nil || response == nil {
+				time.Sleep(backoff)
+				continue
+			}
+
+			expectedParentID := c.lastBlock.GetParent()
+			for _, raw := range response.Blocks {
+				block, err := c.parser.ParseBlock(ctx, raw)
+				if err != nil {
+					continue
+				}
+
+				if expectedParentID != block.GetID() {
+					continue
+				}
+				expectedParentID = block.GetParent()
```

If we receive an invalid entry within `response.Blocks` (fails to parse or unexpected blockID), then we should break rather than continue here because a correct client will never provide us invalid data. In other words, if we hit invalid data, we should assume the rest of the response does not provide helpful data.

---

Right, good catch missed this..

https://github.com/ava-labs/hypersdk/blob/fc7747ed87b6fb9e90ad6f1f48a8dc1ae8dbdc40/internal/validitywindow/client.go#L93